mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala

```diff
@@ -156,6 +156,26 @@ class KMeans private (
     this
   }
 
+  // Initial cluster centers can be provided as a KMeansModel object rather than using the
+  // random or k-means|| initializationMode
+  private var initialModel: Option[KMeansModel] = None
+
+  /** Set the initial starting point, bypassing the random initialization or k-means||
+    * The condition (model.k == this.k) must be met; failure will result in an
+    * IllegalArgumentException.
+    */
+  def setInitialModel(model: KMeansModel): this.type = {
+    if (model.k == k) {
+      initialModel = Some(model)
+    } else {
+      throw new IllegalArgumentException("mismatched cluster count (model.k != k)")
+    }
+    this
+  }
+
+  /** Return the user supplied initial KMeansModel, if supplied */
+  def getInitialModel: Option[KMeansModel] = initialModel
```

Comment:
Does this need to be exposed back to the caller? might be better to not promise this API